CesiumGeometry/include/CesiumGeometry/IntersectionTests.h

```diff
@@ -94,6 +97,129 @@ class CESIUMGEOMETRY_API IntersectionTests final {
       const glm::dvec3& triangleVertB,
       const glm::dvec3& triangleVertC,
       glm::dvec3& barycentricCoordinates) noexcept;
+
+  /**
+   * @brief Tests if a ray hits a triangle and returns the hit point.
+   *
+   * @param ray The ray.
+   * @param p0 The first vertex of the triangle.
+   * @param p1 The second vertex of the triangle.
+   * @param p2 The third vertex of the triangle.
+   * @param cullBackFaces Ignore triangles that face away from ray.
+   * @return The point of intersection, or `std::nullopt` if there is no
+   * intersection.
+   */
+  static std::optional<glm::dvec3> rayTriangle(
+      const Ray& ray,
+      const glm::dvec3& p0,
+      const glm::dvec3& p1,
+      const glm::dvec3& p2,
+      bool cullBackFaces = false);
+
+  /**
+   * @brief Tests if an infinite ray hits a triangle and returns the parametric
+   * hit position.
+   *
+   * The parameter t is positive if the intersection point is in front of the
+   * ray origin, negative if it is behind it, or zero if the two points
+   * coincide.
+   *
+   * @param ray The ray.
+   * @param p0 The first vertex of the triangle.
+   * @param p1 The second vertex of the triangle.
+   * @param p2 The third vertex of the triangle.
+   * @param t Returned parametric value, if a hit is detected.
+   * @param cullBackFaces Ignore triangles that face away from ray.
+   * @return True if a hit occurred, false otherwise.
+   */
+  static bool rayTriangleParametric(
+      const Ray& ray,
+      const glm::dvec3& p0,
+      const glm::dvec3& p1,
+      const glm::dvec3& p2,
+      double& t,
+      bool cullBackFaces = false);
```

Comment:
`rayTriangle` returns an optional, while `rayTriangleParametric` returns a boolean plus an out parameter. Is there a good reason for this difference? The optional seems like the cleaner interface to me, unless we know it to have a significant performance impact (I know this function can be quite performance critical).

Reply:
`rayTriangle` is definitely cleaner.

`rayTriangleParametric` is a little different, in that it's testing an infinite ray. It will return true (hit), even if the hit is behind the ray origin. The caller can interpret this as finite or not, which is what `rayTriangle` is doing

Reply:
Yeah I understand the difference between the two functions as far as an infinite versus finite ray, and returning the intersect distance (which may be negative) instead of the intersection position. But I'm not sure that justifies the different API here. Couldn't `rayTriangleParametric` be equivalently written as:

```cpp
static std::optional<double> rayTriangleParametric(
      const Ray& ray,
      const glm::dvec3& p0,
      const glm::dvec3& p1,
      const glm::dvec3& p2,
      bool cullBackFaces = false);
```

There's no need to impose the hassle of the out parameter on callers, is there? Unless it makes a measurable performance difference in th tight loops where this function is called. It's not obvious to me that would be the case, but I wouldn't be shocked either.

Reply:
Done. Change committed